### ISSUE TITLE
<release>  has no deployed releases

### ISSUE THREAD

**Issue by jkroepke** [databus23/helm-diff#108]
Related: https://github.com/helm/helm/pull/3597

> When using helm upgrade --install, if the first release fails, Helm will respond with an error saying that it cannot upgrade from an unknown state.

The current workaround is to use `helm upgrade --install --force` but a handling like this is missing in helm diff.

```
helm diff upgrade dyn . --allow-unreleased -q 
"dyn" has no deployed releases
Error: "dyn" has no deployed releases
Error: plugin "diff" exited with error
```

It would be great if this plugin could be catch this error.

**Comment by Keramblock** [comment]
Has same issue, when I tried to use this plugin in CI/CD =(

**Comment by mumoshu** [comment]
Yeah this is very frustrating! But my impression was that this should be handled outside of helm-diff.

For example, the latest version of helm added `--atomic` option to `upgrade`, which automatically removes the stale release that causes this issue. Even with the feature, do we still need this handled in helm-diff?

**Comment by jkroepke** [comment]
Poke

Am 08.11.2019 um 06:03 schrieb stale[bot] <notifications@github.com>:

﻿

This issue has been automatically marked as stale because it has not had
recent activity. It will be closed if no further activity occurs. Thank you
for your contributions.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/databus23/helm-diff/issues/108?email_source=notifications&email_token=AAL5AC7Z5SAUBIPG4NSGR2TQSTXINA5CNFSM4GBNKEMKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDOXVRQ#issuecomment-551385798>,
or unsubscribe

**Comment by mumoshu** [comment]
@jkroepke Hey! I'm curious - what helm-diff should do? I'd rather suggest you run `helm upgrade --install --atomic` whenever you install, so that you and `helm-diff` would never need  to handle it.

**Comment by jkroepke** [comment]
>  Hey! I'm curious - what helm-diff should do?

If there is `no deployed releases` handle it like an install.

**Comment by mumoshu** [comment]
@jkroepke Thanks! Then you or your CI run `helm upgrade --install --force`?

**Comment by jkroepke** [comment]
> Thanks! Then you or your CI always run helm upgrade --install --force?

In dev environments and certain changes requires force flag.

But how is the question related to this issue?

**Comment by jkroepke** [comment]
no stale

**Comment by MFAshby** [comment]
I've run into the same issue. `helm diff` already has a `--allow-unreleased` option, I think `upgrade --install` could behave exactly the same (basically an alias for the same command)

FWIW I'm implementing CD and I want a diff of changes to be produced (so someone can review it) when someone raises a PR that changes deployment. Having this change means my script for deploy can be almost identical to my script for checks, so, CI runs:
```
helm diff upgrade --install myapp ./myapp
```
then when the change is merged, CD runs:
```
helm upgrade --install myapp ./myapp
```

I'll make a PR in case this is an acceptable idea.

**Comment by mumoshu** [comment]
@MFAshby Hey! I think your change is fine. But does it really solve the original issue reported by @jkroepke?

**Comment by mumoshu** [comment]
@jkroepke Would you mind submitting a PR for this? If I understand correctly, this should be a matter of adding some error handling logic so that `helm diff` can exit with a successful exit status(0) when the error message is `Error: "RELEASE_NAME" has no deployed releases`.

**Comment by jkroepke** [comment]
I'm don't have any golang skills. 😞

**Comment by MFAshby** [comment]
@mumoshu good point, it does not. I will see if I can include this. I think to match the behaviour of helm, it should only catch the error if `--force` parameter is specified as well? Otherwise `helm diff upgrade` command would succeed but the corresponding `helm upgrade` would fail.

**Comment by MFAshby** [comment]
@jkroepke go is pretty easy to get going with 😉

**Comment by iAnomaly** [comment]
@mumoshu I wanted to add some more color here. Indeed Helm's `--atomic` covers the case where the install/upgrade fails _from Helm's perspective_ and Helm is able to rollback the change.

However, we have an interesting case where our CI/CD runner sometimes dies while Helm upgrade is running. The install/upgrade never completes (the release is stuck in the `PENDING_INSTALL` state) and `--atomic` logic never gets a chance to execute and cleanup (because Helm process died along with the runner).

I'm going to review the Helm diff code, but I'm wondering if `--allow-unreleased`'s behavior could be expanded to also treat any release in the `PENDING_INSTALL` state as "unreleased".

**Comment by mumoshu** [comment]
@iAnomaly Good point! I believe you're correct. Expanding `--allow-unreleased`'s behavior sounds great.

Also - Maybe helm-diff could optionally return a non-zero exit code like 2 or 3(perhaps 3 might be better as we already use 2 for `-diff --detailed-exitcode`) so that the later process can choose to add `--force` flag to `helm upgrade --install`.

**Comment by jkroepke** [comment]
no stale

**Comment by dody87** [comment]
no stale. Same issue here. No solutions yet.

**Comment by jkroepke** [comment]
I would appreciate it to label this issues with `pinned` to exclude them from stale bot.

**Comment by jkroepke** [comment]
you won.

**Comment by mumoshu** [comment]
@jkroepke I've already suggested possible solutions but anyone has submitted a PR for any of those yet. What does "won" mean in this context? Who can you win? By forcing anyone to fix the issue for you?

**Comment by jkroepke** [comment]
> What does "won" mean in this context?

The bot want to close this issue, I just do it for him.

In general, ignore the exit code of helm diff should be best option here. Thats the solution for me here.
